src/Sarif.Multitool/Rules/SARIF1021.MissingToolVersion.cs

```diff
@@ -0,0 +1,40 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using Microsoft.Json.Pointer;
+
+namespace Microsoft.CodeAnalysis.Sarif.Multitool.Rules
+{
+    public class MissingToolVersion : SarifValidationSkimmerBase
+    {
+        private readonly MultiformatMessageString _fullDescription = new MultiformatMessageString
+        {
+            Text = RuleResources.SARIF1021_MissingToolVersion
+        };
+
+        public override MultiformatMessageString FullDescription => _fullDescription;
+
+        public override FailureLevel DefaultLevel => FailureLevel.Warning;
+
+        /// <summary>
+        /// SARIF1021
+        /// </summary>
+        public override string Id => RuleId.MissingToolVersion;
+
+        protected override IEnumerable<string> MessageResourceNames => new string[]
+        {
+            nameof(RuleResources.SARIF1021_Default)
+        };
+
+        protected override void Analyze(Tool tool, string pointer)
+        {
+            // 'Driver' is a required property, hence we do not need a null check for it. 
+            if (string.IsNullOrWhiteSpace(tool.Driver.Version))
+            {
+                LogResult(pointer.AtProperty(SarifPropertyName.Driver), nameof(RuleResources.SARIF1021_MissingToolVersion), tool.Driver.Name);
```

Comment:
>LogResult [](start = 16, length = 9)

It is very helpful to recapitulate the user-facing string as a code comment, because it makes it quite clear what's happening in this return condition. Also, I think LogResult should *require* an argument that indicates the failure severity. That information is also important for rules maintainability. See this example rule from BinSkim that demonstrates the standard: [https://github.com/microsoft/binskim/blob/master/src/BinSkim.Rules/PERules/BA2010.DoNotMarkImportsSectionAsExecutable.cs](https://github.com/microsoft/binskim/blob/master/src/BinSkim.Rules/PERules/BA2010.DoNotMarkImportsSectionAsExecutable.cs)